### ISSUE TITLE
"Reveal path to evacuation center" does not follow roads.

### ISSUE THREAD

**Issue by kevingranade** [CleverRaven/Cataclysm-DDA#20204]
When triggering the terminal in the starting shelter, the revealed path leads through forests, swamps, etc, rendering it useless.

To reproduce:
Spawn character, interact with terminal in NE corner of shelter, select option 3.
Open overmap, note that revealed route goes through forests.

Expected outcome:
Revealed route follows roads or the console prints it's error message about not being able to find a route to the shelter via road.

This is a key intro to the Evac Center questlines, so I'm marking it as a blocker for 0.D.

The simplest fix that has been proposed is:
> doing as it is now, and returning false from overmapbuffer::reveal_route() when res is over some cutoff.

In other words, detecting if it isn't a road-only path and failing when that happens.

If the A* code supports it, it would be best to only have it consider road tiles in the first place, this would make its working set much smaller and might make it dramatically faster.

**Comment by Zireael07** [comment]
I think the path is just an A* search, so it looks for the shortest route. I don't understand why being shorter renders it useless?

**Comment by keyspace** [comment]
Does this happen even if the evac shelter is on the same overmap?..

**Comment by CoroNaut** [comment]
If it reveals the path using roads then one can assume the computer would have the information for much more than just the paths specific tiles. Shouldn't it reveal gas stations, towns, FEMA camps and other things that would be useful for an evacuee?

**Comment by keyspace** [comment]
@CoroNaut IMO that's a separate issue - expanding the console's functionality.

**Comment by patternoia** [comment]
[Here ](https://github.com/CleverRaven/Cataclysm-DDA/pull/19688#issuecomment-266245906) codemime speaks about shortucts as intended behavior:
>Yes, when the algorithm cannot find a regular path (poorly connected roads), it sometimes takes shortcuts through the wilderness.

**Comment by kevingranade** [comment]
The only reason to follow roads at all is to make the path navigable for
vehicles, if that's not insured, you might as well draw a straight line to
the destination or just set the coordinates in a mission  (which is already
happening).

**Comment by kevingranade** [comment]
On Feb 6, 2017 8:33 AM, "patternoia" <notifications@github.com> wrote:

Here
<https://github.com/CleverRaven/Cataclysm-DDA/pull/19688#issuecomment-266245906>
codemime speaks about shortucts as intended behavior:

Yes, when the algorithm cannot find a regular path (poorly connected
roads), it sometimes takes shortcuts through the wilderness.


That's a known issue, not intended behavior.  Also these aren't shortcuts,
it looks like it just gives up and does an x or you intercept at some point.

**Comment by keyspace** [comment]
> Also these aren't shortcuts, it looks like it just gives up and does an x

I've only seen this happen when the refugee center in on a different overmap. If it's on the same one, the road traversal algorithm seems perfectly fine with going as long as needed.

Haven't looked at the code, but guess that if the closest refugee center is on another overmap, the algorithm just gets as close as it can via roads to the other overmap, and then cuts straight through. Not sure if it starts following the road if it "reacquires" one after losing it once.

**Comment by keyspace** [comment]
In `mission_start::reveal_refugee_center()`:

``` c++
1744     const tripoint source_road = overmap_buffer.find_closest( your_pos, "road", 3, false );
1745     const tripoint dest_road = overmap_buffer.find_closest( center_pos, "road", 3, false );
1746 
1747     if( overmap_buffer.reveal_route( source_road, dest_road ) ) {
1748         add_msg( _( "You mark the refugee center and the road that leads to it..." ) );
1749     } else {
1750         add_msg( _( "You mark the refugee center, but you have no idea how to get there by road..." ) );
1751     }
```

The "have no idea" message is only printed when `overmapbuffer::reveal_route()` returns `false` - i.e. when the `path` found by `pf::find_path()` is `empty()`.

There doesn't seem to be many causes for that to happen (in `simple_pathfinding.h`), and they mostly look like "failure, bailing, here's an empty vector" kinds.

**Comment by patternoia** [comment]
>here doesn't seem to be many causes for that to happen
On the moment when I first introduced this, it was the case, so I have covered it with appropriate message.

**Comment by keyspace** [comment]
@patternoia It works as expected. It uses an `estimate` predicate passed from `overmapbuffer::reveal_route()`, and that [gives a heavy penalty](https://github.com/CleverRaven/Cataclysm-DDA/blob/9f3f8b3/src/overmapbuffer.cpp#L489) to routes that go through non-road overmap tiles.

However, if there is _no route at all_, then a very high-cost one will be considered valid anyway - `estimate` never returns negative.

Possible solutions depend on what's desired here:

* find a road route, even if it takes forever, by increasing radii all around - very dependent on hardware, so not fool-proof;
* allow returning no route at all, even if the best one only had a short walk through the swamp - can be achieved by returning negative from `estimate`;
* doing as it is now, and returning `false` from `overmapbuffer::reveal_route()` when `res` is over some cutoff.

I think the latter is best.

**Comment by kevingranade** [comment]
On Feb 6, 2017 3:02 PM, "Keyspace-1" <notifications@github.com> wrote:

@patternoia <https://github.com/patternoia> It works as expected. It uses
an estimate predicate passed from overmapbuffer::reveal_route(), and that gives
a heavy penalty
<https://github.com/CleverRaven/Cataclysm-DDA/blob/9f3f8b3/src/overmapbuffer.cpp#L489>
to routes that go through non-road overmap tiles.

However, if there is *no route at all*, then a very high-cost one will be
considered valid anyway - estimate never returns negative.

Possible solutions depend on what's desired here:

   - find a road route, even if it takes forever, by increasing radii all
   around - very dependent on hardware, so not fool-proof;
   - allow returning no route at all, even if the best one only had a short
   walk through the swamp - can be achieved by returning negative from
   estimate;
   - doing as it is now, and returning false from
   overmapbuffer::reveal_route() when res is over some cutoff.

Does this mean, "give up and reveal nothing if we couldn't find a road-only
path"?  It could stand to try a bit harder to find a route, including
making shelters more common or even roads more connected, but no path is
better than one that doesn't follow roads.

**Comment by keyspace** [comment]
There is no way to guarantee a route unless we guarantee that every overmap will be connected by road to all four of its neighbouring overmaps.

Otherwise, there is always a chance, even if very small, that:

* the closest refugee center is on an adjacent overmap, but
* the best road-only route to it leads through 42 other overmaps.

The current radius of search is [4 overmaps](https://github.com/CleverRaven/Cataclysm-DDA/blob/9f3f8b3/src/overmapbuffer.cpp#L468), and that seems large enough to me.

If there's no road to the refugee center, the player will have to walk an overmap border and find another shelter with a console there.

**Comment by keyspace** [comment]
One more thing: when there _is_ a road-only route from one point to another, the current "walker" considers [absolute distance between `start` and `finish` points](https://github.com/CleverRaven/Cataclysm-DDA/blob/9f3f8b3/src/overmapbuffer.cpp#L492) as the route's weight.

It seems more intuitive to me to add a tiny weight to each tile travelled instead.

I also haven't reviewed if the function is used by any other missions, where "road-onlyness" may be unachievable.

**Comment by kevingranade** [comment]
> There is no way to guarantee a route unless we guarantee that every overmap will be connected by road to all four of its neighbouring overmaps.

This doesn't strike me as an unreasonable guarantee to make in general.  If we ever add wilderness only over maps or similar, that could suppress it, but otherwise adjacent towns or other destinations would tend to be connected.

**Comment by svick** [comment]
What I as a player would like to see is if it showed a route that can be traversed by a car. That does not mean it has to follow roads, but it does mean it can't contain any forests/swamps/whatever.

Put another way, I would like to see the fastest route, not shortest, where the cost is:

* roads have normal cost
* fields (and other tiles traversable by car) have some higher cost
* forests (and other tiles that are not traversable) are not considered at all

Though I have no idea how hard would it be to implement or of this solution has other issues.

**Comment by keyspace** [comment]
@svick:

> What I as a player would like to see is if it showed a route that can be traversed by a car.

Whitelisting/blacklisting in code is trivial and [already done now](https://github.com/CleverRaven/Cataclysm-DDA/blob/9f3f8b3/src/overmapbuffer.cpp#L483).

However, It'd be preferable to avoid such hard-coding in `overmap_buffer.reveal_route()`, since

> the function is used by [...] other missions, where "road-onlyness" may be unachievable.

What would be nice is an optional "allowed terrains" predicate that could be used instead of the current "default and only one". JSON might also need a new field then - currently only `see_cost` might be usable, and that's not its intended purpose.

The above sounds like too much refactoring for the purposes of `0.D` - IMO best delayed.

-----

> > There is no way to guarantee a route unless we guarantee that every overmap will be connected by road to all four of its neighbouring overmaps.

> This doesn't strike me as an unreasonable guarantee to make in general. If we ever add wilderness only over maps or similar, that could suppress it,

OK, sounds like there's a candidate solution.

**Comment by codemime** [comment]
> One more thing: when there is a road-only route from one point to another, the current "walker" considers absolute distance between start and finish points as the route's weight. It seems more intuitive to me to add a tiny weight to each tile travelled instead.

These are completely different things. The absolute distance encourages going towards the destination. I mean, it's a matter of direction which has nothing to do with the distance actually travelled.

What about craters on roads? The map shouldn't know a thing about them, but they already exist and have heavy impact on pathfinding. They're good candidates to becoming overlays (being revealed only by seeing, see #19387), but how do we treat them till then?

**Comment by codemime** [comment]
> this would make its working set much smaller and might make it dramatically faster.

The bottleneck here isn't the pathfinding (which is fast enough already), but finding the actual destination for it. That's done using a "tile by tile" brute force approach (sloooow).

**Comment by Zireael07** [comment]
> The bottleneck here isn't the pathfinding (which is fast enough already), but finding the actual destination for it. That's done using a "tile by tile" brute force approach (sloooow).

Could perhaps be sped up if the specials creation in mapgen returned the coords? Then you could map:getSpecialLocation("refugee center") or some such and avoid the slooow tile by tile search.

**Comment by kevingranade** [comment]
On Feb 12, 2017 9:07 AM, "codemime" <notifications@github.com> wrote:

this would make its working set much smaller and might make it dramatically
faster.

The bottleneck here isn't the pathfinding (which is fast enough already),
but finding the actual destination for it. That's done using a "tile by
tile" brute force approach (sloooow).

Oh, we can trivially cache specials in the overmap when they're generated
or loaded if that's a problem.

**Comment by Coolthulhu** [comment]
Cached specials could be useful, but why is brute find slow? It should be just flood fill.
Is it the cost of global coord->overmap->overmap coord->tile check?

**Comment by Coolthulhu** [comment]
In #20402 I disabled the non-road paths. Is that enough?

**Comment by kevingranade** [comment]
Totally, anything else is icing.

**Comment by Coolthulhu** [comment]
Is it "in progress" because it still needs a better fix?

**Comment by kevingranade** [comment]
No, just because I haven't updated it, and I can't do so from by phone, bah.
